lib/ElementHandle.js

```diff
@@ -137,24 +151,72 @@ class ElementHandle extends JSHandle {
     return {x, y, width, height};
   }
 
+  /**
+   * @return {!Promise<?{x: number, y: number, width: number, height: number}>}
+   */
+  async _assertBoundingBox() {
+    const boundingBox = await this.boundingBox();
+    if (!boundingBox)
+      throw new Error('Node is not visible');
+    return boundingBox;
+  }
+
   /**
    *
    * @param {!Object=} options
    * @returns {!Promise<Object>}
    */
   async screenshot(options = {}) {
-    await this._scrollIntoViewIfNeeded();
+    let needsViewportReset = false;
+
+    let boundingBox = await this._assertBoundingBox();
+
+    const viewport = Object.assign({}, this._page.viewport());
+
+    const newViewport = {
+      width: Math.max(viewport.width, Math.ceil(boundingBox.width)),
+      height: Math.max(viewport.height, Math.ceil(boundingBox.height)),
+    };
+
+    const initialScroll = await this.executionContext().evaluate(function() {
+      return { scrollX: window.scrollX, scrollY: window.scrollY };
+    });
+
+    if (newViewport.width !== viewport.width || newViewport.height !== viewport.height) {
+      await this._page._emulationManager.emulateViewport(this._client, newViewport);
+
+      needsViewportReset = true;
+    }
+
+    await this.executionContext().evaluate(function(el) {
+      const box = el.getBoundingClientRect();
+
+      window.scrollTo(
```

Comment:
I'm not understanding how scrollIntoViewIfNeeded fails here. I can see how you might need to call it again after resizing the viewport.

Reply:
This is with `await this._scrollIntoViewIfNeeded();`. Part of the element and its right border get cut off.

![screenshot-element-larger-than-viewport](https://user-images.githubusercontent.com/936069/34855493-3bffc1c0-f6fd-11e7-9715-146e4cdd4a65.png)

Reply:
It looks like the viewport isn’t wide enough though? The left side of of the image is flush against the left of the viewport.

Reply:
There's a 50px left margin. The only difference between the full and right-cropped version is the scroll logic. The viewport size is exactly the same. It's not actually right cropped, more like clipped. There's some white space on the right side of the image. I think that part of the element is outside `<body>`.

Could probably get it working with `scrollIntoViewIfNeeded` if we set the viewport width minimum to be `bounds.left + bounds.width`. That has the cost of increasing our chance of resizing the browser, and gives an extra thing that can be hit by media queries and percentages.

Reply:
Sorry for the delay. I understand now!

You can use code like this instead of `window.scrollTo`:
```js
await this.executionContext().evaluate(el => {
  el.scrollIntoView({
    behavior: 'instant',
    block: 'center',
    inline: 'nearest'
  });
}, this);
```

`window.scrollTo` doesn't work because the element that needs to scroll might not be the window.

Reply:
Tried that. Test failed with this output:

![screenshot-element-larger-than-viewport](https://user-images.githubusercontent.com/936069/35247266-01975902-ff87-11e7-80fe-d874b3cd29cf.png)

You're not going to like this, but it seems to be working with this code, tested against two scroll parent divs plus window.

```js
    await this.executionContext().evaluate(function(el) {
      const parents = [];
      for (let parent = el.parentNode; parent; parent = parent.parentNode) {
        if (parent.scrollHeight !== parent.offsetHeight)
          parents.unshift(parent);
      }

      for (const parent of parents) {
        parent.scrollTo(0, 0);

        const parentBounds = parent.getBoundingClientRect();
        const elBounds = el.getBoundingClientRect();

        const x = elBounds.left - parentBounds.left;
        const y = elBounds.top - parentBounds.top;

        parent.scrollTo(x, y);
      }

      el.scrollIntoView({
        behavior: 'instant',
        block: 'center',
        inline: 'nearest',
      });
    }, this);
```

Reply:
This could be optimized, but I'll do that if you don't give a hard **no** 😄

Reply:
I'd like to avoid it. I'll give it a shot later. `scrollIntoView` passed the test for me when I tried it. What OS are you on?

Reply:
OSX. The test is `should screenshot element with scroll container`.